### PR TITLE
Ring Hash LB Policy Implementation

### PR DESCRIPTION
**Pull request by donnadionne**
We need to increase the error tolerance slightly for the ring hash
tests; increasing only the tolerance will cause less RPCs to be sent via
ComputeIdealNumRpcs method which assumes normal distribution and uses the
99.99% standard diviation.

It is likely the ring hash distribution is not the normal distribution
and may have higher standard deviation.

Adding a parameter to ComputeIdealNumRpcs to allow standard deviation
value to be passed in and allow a higher computed RPC number to be
calculated.

Retested all flaky tests 1000x and no longer see the flake.

### CODE REVIEW

**Comment by markdroth** [comment]
Why is the distribution not random?  I think it should be.  My guess is that what we're seeing here is a real bug uncovered by the test in the way we're computing the random hash.

Right now, we're generating a 32-bit random number with `rand()`, writing it to a string, and then passing it to the hash function.  But that may not be the right approach; the hash function probably won't necessarily distribute a bunch of different 32-bit numeric strings to random parts of the 64-bit hash space.  Instead, how about we call `rand()` twice, once for the high-order 32 bits and then again for the low-order 32 bits?

File: test/cpp/end2end/xds_end2end_test.cc [hunk 1]
```diff
@@ -6858,17 +6858,22 @@ TEST_P(CdsTest, RingHashNoHashPolicy) {
   gpr_setenv("GRPC_XDS_EXPERIMENTAL_ENABLE_RING_HASH", "true");
   const double kDistribution50Percent = 0.5;
   const double kErrorTolerance = 0.05;
+  const uint32_t kRpcTimeoutMs = 5000;
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:6861]: Why does increasing the ring size require increasing the timeouts?  The two endpoints are both up, so neither of them should cause delays from trying to connect.

Same thing for all of these tests.

**Comment by markdroth** [comment]
Also, in addition to these test changes, I think we should also change the XdsConfigSelector code to generate the random hash by calling `rand()` twice, once for the high-order 32 bits and then again for the low-order 32 bits.  I think that is both more efficient than what we're doing now and will yield better randomness.

File: test/cpp/end2end/xds_end2end_test.cc (hunk 1 above)
Comment (donnadionne) [reply test/cpp/end2end/xds_end2end_test.cc:6861]: It takes seconds to create the Picker for ring hash policy, and we do that multiple times: idle->connecting, connecting->ready, and for multiple channel changes
I0517 06:37:57.427480420      20 ring_hash.cc:343]           donna starting to create the picker: the ring
I0517 06:37:59.829846291      20 ring_hash.cc:376]           [RH 0x7b14000014f0 picker 0x7b0c00017e20] created picker from subchannel_list=0x7b7400012200 with 100000 ring entries

So the very first RPC can timeout

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc:6861]: Woah -- that's surprising.  Why is it taking so long to create the picker?

As evidenced by this test, we create new pickers whenever any one individual subchannel changes state, and that can happen pretty frequently, so picker creation really needs to be fast.  If it's taking this long, then we need to find a way to optimize it.

**Pull request closed without merging**